packages/dso-toolkit/gulp/build-svg-spritesheets.js

```diff
@@ -0,0 +1,165 @@
+/* eslint-env node */
+
+import gulp from "gulp";
+import { basename } from "path";
+import { obj } from "list-stream";
+import { parse } from "css";
+import svgstore from "gulp-svgstore";
+import cheerio from "gulp-cheerio";
+import prettier from "gulp-prettier";
+import rename from "gulp-rename";
+
+import { sassTransformer } from "./transformers/sass.transformer.js";
+
+const iconsPath = "src/icons";
+const distPath = "dist";
+const outPutNameIcons = "dso-icons.svg";
+const outPutNameIconsWithVariants = "di.svg";
+const canvasSize = 32;
+const gutterSize = 10;
+
+async function parseStylesheets() {
+  return new Promise((resolve, reject) => {
+    gulp
+      .src(`${iconsPath}/*.scss`) // Read all SCSS files in the icons directory
+      .pipe(sassTransformer()) // Transform SCSS into plain CSS
+      .pipe(
+        obj((error, data) => {
+          if (error) {
+            return reject(error);
+          }
+
+          // Process each transformed SCSS file
+          const parsedStyles = data.map(({ relative, extname, contents }) => {
+            const id = basename(relative, extname); // Extract the file name without the extension
+            const style = contents.toString();
+            const ast = parse(style); // Parse the CSS into an abstract syntax tree (AST)
+
+            // Extract color variants for each CSS rule
+            const variants = ast.stylesheet.rules
+              .filter((rule) => rule.type === "rule") // Keep only CSS rules (ignore comments, etc.)
+              .flatMap((rule) => {
+                const colorDeclaration = rule.declarations.find((d) => d.property === "color");
+                if (!colorDeclaration) return [];
+
+                const color = colorDeclaration.value;
+                return rule.selectors
+                  .map((selector) => {
+                    if (selector === id) {
+                      return { selector: "default", color }; // Mark as default variant
+                    }
+
+                    if (selector.startsWith(`${id}:`)) {
+                      return { selector: selector.slice(id.length + 1), color }; // Extract the variant name
+                    }
+
+                    return null;
+                  })
+                  .filter(Boolean); // Remove null entries
```

Comment:
Een `.map()` en `.filter()` kan samen in `.reduce()`.

Dat scheelt een extra iteratie en een nieuwe array.

Leesbaarheid staat hier altijd weer ter discussie, wat leest duidelijker? Eerst een map naar nieuw format en daarna een filter, of de filter logica verwerken in de mapping logica.

Mijn overtuiging in deze is dat je met minder, meer bereikt: Er wordt nu `null` geintroduceerd die er later uitgefilterd wordt. Er staat zelfs een begeleidende comment bij. De `null` kan alleen in context van de voorgaande mapping worden begrepen.